xtra-libp2p-offer/src/current/maker.rs

```diff
@@ -0,0 +1,139 @@
+                xtra_libp2p::NegotiationError::Failed,
+            )) = e.downcast_ref::<xtra_libp2p::Error>()
+            {
+                // It's normal to disagree on the protocols now that we broadcast on both versions
+                // to _all_ our peers
+            } else {
+                tracing::warn!(%peer_id, "Failed to send offers: {e:#}")
+            }
+        };
+
+        let this = ctx.address().expect("self to be alive");
+        spawn_fallible(&this, task.instrument(span), err_handler);
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(&mut self, msg: NewOffers, ctx: &mut xtra::Context<Self>) {
+        self.latest_offers = msg.0;
+
+        let quiet = quiet_spans::sometimes_quiet_children();
+        for peer_id in self.connected_peers.iter().copied() {
+            self.send_offers(peer_id, ctx)
+                .instrument(quiet.in_scope(|| {
+                    tracing::debug_span!("Broadcast offers to taker (libp2p)").or_current()
+                }))
+                .await
+        }
+    }
+
+    async fn handle(&mut self, _: GetLatestOffers) -> Vec<model::Offer> {
+        self.latest_offers.clone()
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle_connection_established(
+        &mut self,
+        msg: endpoint::ConnectionEstablished,
+        ctx: &mut xtra::Context<Self>,
+    ) {
+        tracing::trace!("Adding newly established connection: {:?}", msg.peer_id);
+        self.connected_peers.insert(msg.peer_id);
+        self.send_offers(msg.peer_id, ctx).await;
+    }
+
+    async fn handle_connection_dropped(&mut self, msg: endpoint::ConnectionDropped) {
+        tracing::trace!("Remove dropped connection: {:?}", msg.peer_id);
+        self.connected_peers.remove(&msg.peer_id);
+    }
+}
+
+/// Instruct the `offer::maker::Actor` to broadcast to all
+/// connected peers an update to the current offers.
+pub struct NewOffers(Vec<model::Offer>);
+
+impl NewOffers {
+    pub fn new(offers: Vec<model::Offer>) -> Self {
+        Self(offers)
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct GetLatestOffers;
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = ();
+
+    #[tracing::instrument(name = "xtra_libp2p_offer::maker::Maker started", skip_all)]
+    async fn started(&mut self, ctx: &mut xtra::Context<Self>) {
+        match self.endpoint.send(GetConnectionStats).await {
+            Ok(connection_stats) => self
+                .connected_peers
+                .extend(connection_stats.connected_peers),
+            Err(e) => {
+                tracing::error!(
+                    "Unable to receive connection stats from the endpoint upon startup: {e:#}"
+                );
+
+                // This code path should not be hit, but in case we run into an error this sleep
+                // prevents a continuous endless loop of restarts.
+                tokio_extras::time::sleep(Duration::from_secs(2)).await;
```

Comment:
Is this referring to the actor restarts? If so, couldn't this be better served by a supervisor restart policy with timeout?

Reply:
I only moved this code around, but unfortunately the diff isn't very good :(

I don't fully understand why we need this, but what you suggest sounds like a great idea!!

Reply:
I've created https://github.com/itchysats/itchysats/issues/2717 to track this!

Reply:
I think this is actually outdated and should be removed because there is already a more global strategy in place. But I think it is a good idea to tackle it outside of this PR. Thanks for adding the ticket!